README.md
# BOTK\Context
[![Build Status](https://img.shields.io/travis/linkeddatacenter/BOTK-context.svg?style=flat-square)](http://travis-ci.org/linkeddatacenter/BOTK-context)
[![Code Coverage](https://img.shields.io/scrutinizer/coverage/g/linkeddatacenter/BOTK-context.svg?style=flat-square)](https://scrutinizer-ci.com/g/linkeddatacenter/BOTK-context)
[![Latest Version](https://img.shields.io/packagist/v/botk/context.svg?style=flat-square)](https://packagist.org/packages/botk/context)
[![Total Downloads](https://img.shields.io/packagist/dt/botk/context.svg?style=flat-square)](https://packagist.org/packages/botk/context)
[![License](https://img.shields.io/packagist/l/botk/context.svg?style=flat-square)](https://packagist.org/packages/botk/context)

Ultraligth validation and sanitization helpers.

This is a BOTK package. Please refer to http://ontology.it/tools/botk for more info
about BOTK project.

## Quick start

The package is available on [Packagist](https://packagist.org/packages/botk/context).
You can install it using [Composer](http://getcomposer.org).

```bash
composer require botk/context
```

Some code examples in samples directory.

## BOTK\Context package documentation


### Abstract

This package exports a set of class to manage the context of a RESTful request.

A Context is defined as the full set of variables available runtime durin the serving of an http request.
          Variables are grouped in distinct name spaces.

All variables in name spaces are **read only**.

This package exposes the methods to sanitize and validate context variables.

Note that cookies and session aren't part of context. This because proper RESTful resource implementation
          shouldn't need persisten status in any way.


### Installation

This package follows[ BOTK guide line for installation](../overview/#installation) and require [composer](http://getcomposer.org/).

Add following dependances to **composer.json** file in your project root:

    {
        "require": {
            "botk/context": "*"
        }
    }

### The Context class

A Context is the full set of all that a CGI endpoint knows about&nbsp; Request beside its URI. Context space
        contains system environment variables, requests header, request body, server specific vars, cookies, variables
        in configuration files, local define variables, etc etc..

Context variables are grouped in _namespaces_.` BOTK\Context\Context` class provide the
        method&nbsp; ` ns(<var>namespace name</var>)` to access name spaces. There are five predefined
        namespaces:

1.  the variables defined in the heading of the request and exposed by web server (<var>INPUT_SERVER</var>); 
3.  the variables encoded in http URI (<var>INPUT_GET</var><var></var>); 
5.  the variables encoded in http body (<var>INPUT_POST</var>);
7.  the system environment (<var>INPUT_ENV</var>); 
9.  local defined variables (Context::LOCAL).


Beside these, Context class allow you to define dynamic namespace built from .ini files loaded runtime from a
configuration directory. The name of the namespace is the name of the configuration file without .ini extension.

The configuration file must exists when is first referenced. The default configuration directory is
        ../configs, that normally is a directory placed at the same level of the directory that contains the scripts who
        instance a context class. You can change the default directory defining the environment variable $_ENV['BOTK_CONFIGDIR'].

Context shold be considered as a readonly data structure because you cant change the resource execution context.

For example suppose to have a end-point script is in ` myapp/httdocs/index.php ` so, to get a
        success on following code, it must exist ` myapp/configs/sample.ini ` file:

    use BOTK\Context\Context as CX;
    $sampleNameSpace = CX::factory()->ns('sample');

If you want to put the .ini file in the same end-point directory:

    use BOTK\Context\Context as CX;
    $_ENV['BOTK_CONFIGDIR'] = '.';
    $sampleNameSpace = CX::factory()->ns('sample');

You can also access to local variable passing them to context constructor:

    $myvar = 'ok'; // define a variable in local scope
    $v = CX::factory(get_defined_vars())->ns(CX::LOCAL)->getValue('myvar');
    //$v == 'ok'

Context class exposes the method `guessRequestCanonicalUri()` that returns the current request uri
        in canonical form and the method `guessRequestRelativelUri()` that returns it as relative path.

Please note that get request uri is not a trivial job, this because http servers and proxyes can change what
        the user entered. You can just guess about it.

### The ContextNameSpace Class

This class implements the controlled access to set of variables.

#### The getValue() method

The `BOTK\Context\ContextNameSpace` class exports the method ` getValue() ` that
        allows to validate and sanitize variables defined in a namespace (see [PHP
          Validation & Sanitization](http://foaa.de/blog/2012/11/27/php-validation-and-sanitization/) article).

If somethings goes wrong it throwns an error that can be trapped by standard BOTK error management.

`getValue()` exposes following interface:

In addition to `getValue()``ContextNameSpace` class expose a set of shortcuts for
        easy to read and write code.

### validator shortcucts

Some validators shold be boring to write, here are a set of predefined shortcuts:

### getValue shortcucts

Even with validator shortcuts, a call to getValue method can be too verbose, here are some shortcut you can use
        for common task:

*  getString($varName,$default = null):&nbsp; same as
              getValue($varName,$default, null, FILTER_SANITIZE_STRING);
*  getURI($varName,$default = null):&nbsp; same as
              getValue($varName,$default, self::STRING('/.+/'), FILTER_SANITIZE_URL)

### PagedResourceContext class

The PagedResource class is a facility to manage the context of a resource whose query string can contains
        variables to drive pagination processing. It is a specialization of Context class and like Context should be
        considered as a read-only data structure.

The resource paging context is inspired on W3C's Linked Data Platform Paging specifications .Here are some imported definitions:
<dl class="glossary"><dt><dfn>Paged resource</dfn></dt><dd>A resource&nbsp; whose representation may be too large to fit in a single HTTP response, for which a server
          offers a sequence of single-page resources. A paged <var>P</var> is broken into a sequence of pages
          (single-page resources) <var>P<sub>1</sub>, P<sub>2</sub>, ...,P<sub>n</sub></var>, the representation of
          each <var>P<sub>i</sub></var> contains a subset of&nbsp; <var>P</var>.. </dd><dt><dfn>Single-page resource</dfn></dt><dd>One of a sequence of related resources <var>P<sub>1</sub>, P<sub>2</sub>, ...,P<sub>n</sub></var>, each of
          which contains a subset of the state of another resource <var>P</var>. <var>P</var> is called the paged
          resource. For readers familiar with paged feeds, a single-page resource is similar to a feed document and the
          same coherency/completeness considerations apply.

Note: the choice of terms was designed to help authors and readers clearly differentiate between the <a title="Paged resource">_resource
                being paged_</a>, and the <a title="Single-page resource">_individual page resources_</a>,
            in cases where both are mentioned in close proximity.

</dd><dt><dfn>first page 
</dfn></dt><dd>The uri to the first <a style="text-decoration: underline;" title="Single-page resource">single-page
            resource</a> of a <a title="Paged resource">paged resource</a><var>P</var>. For example
          http://www.example.org/bigresource?page=0 </dd><dt><dfn>next page 
</dfn></dt><dd>The uri&nbsp; to the next <a title="Single-page resource">single-page resource</a> of a <a title="Paged resource">paged
            resource</a><var>P</var>. </dd><dt><dfn>last page
</dfn></dt><dd>The uri&nbsp; to the last <a title="Single-page resource">single-page resource</a> of a <a title="Paged resource">paged
            resource</a><var>P</var>. </dd><dt><dfn>previous page 
</dfn></dt><dd>The uri&nbsp; to the previous <a title="Single-page resource">single-page resource</a> of a <a title="Paged resource">paged
            resource</a><var>P</var>. </dd></dl>

Pagination require some variable to be specified in resource URI query strings. Such variables name and other
        default values can be passed to class constructor as an option associative array. Here are supported keywords:

*   plabel : contains the variable that contains page num. For default 'page'
*   pslabel&nbsp; : contains the variable that contains page size. For default 'pagesize
*   pagesize&nbsp; : contains the default value for page size. The default is 100

It exposes following methods:

<dl><dt>`getSinglePageResourceUri( $pagenum = null, $pagesize=null)`</dt><dd>returns the canonical uri with page variables in query string. Without arguments returns current page
          resource uri</dd><dt>`getPagedResourceUri()`</dt><dd>return the resource uri wit page variables removed from query string&nbsp; </dd><dt>`isPagedResource()`</dt><dd>returns true is resource uri contains page info</dd><dt>`getPageNum()`</dt><dd>returns the current page num</dd><dt>`getPageSize()`</dt><dd>returns the size of the page</dd><dt>`firstPageUri()`</dt><dd>returns the Resource uri for first Single page resource</dd><dt>`nextPageUri()`</dt><dd>returns the Resource uri for next page resource or the current page resource uri</dd><dt>`prevPageUri()`</dt><dd>returns the Resource uri for next page resource or the first page resource uri</dd><dt>`hasNextPage()`</dt><dd>returns true if a next page is available. By default returns true. It can be affected by `declareActualPageSize()`
          method</dd><dt>`isLastPage()`</dt><dd>returns true if the current is the last page. By default returns false. It can be affected by `declareActualPageSize()`
          method</dd><dt>`declareActualPageSize($count)`</dt><dd>inform context of the number of item in current Page, by default is 0\. This affects `hasNextPage()`
          and `isLastPageMethods()`.</dd></dl>

Here is how `declaretActualPageSize()` affects `hasNextPage()` and `isLastPage()`:

## License

 Copyright © 2016 by  Enrico Fagnoni at [LinkedData.Center](http://LinkedData.Center/)®

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


  

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/) and [changelog format](http://keepachangelog.com/)

## [Unreleased]

## [4.1.0] - 2016-02-10
### Addedd
-  NULL_AS_DEFAULT support

## [4.0.3] - 2016-01-29
### Changed
- some project cleaning


## [4.0.2] - 2016-01-28
### Added
- vagrant support
- contribution instructions
- badge management
- trevis-ci support
- scrutinizer support
### Changed
- doc translated in markdown

## [4.0.1] - 2016-01-28
### Added
- integration with packagist
- changelog alignement


## 4.0.0 - 2016-01-28
### Added
- Moved from google code
- Code, doc and tests completed

[Unreleased]:  https://github.com/linkeddatacenter/BOTK-context/compare/4.1.0...HEAD
[4.1.0]:  https://github.com/linkeddatacenter/BOTK-context/compare/4.0.3...4.1.0
[4.0.3]:  https://github.com/linkeddatacenter/BOTK-context/compare/4.0.2...4.0.3
[4.0.2]:  https://github.com/linkeddatacenter/BOTK-context/compare/4.0.1...4.0.2
[4.0.1]:  https://github.com/linkeddatacenter/BOTK-context/compare/4.0.0...4.0.1


test/_support/install.sh
#!/usr/bin/env bash

PROJECT="BOTK-context"
REPO="https://github.com/linkeddatacenter/BOTK-context.git"
WEB_ROOT="/opt/$PROJECT/samples"

###########################################################################
# This scripts install all needed component in a fresh ubuntu 14.04 server
###########################################################################
if [ ! -f /tmp/install.lock ]; then
	apt-get update
	echo "apt-get update done" > /tmp/install.lock
	apt-get -y install git apache2 php5-common php5-curl libapache2-mod-php5 php5-cli curl phpunit

	curl -sS https://getcomposer.org/installer | sudo php -- --install-dir=/usr/local/bin --filename=composer

	# Clone  project in /opt (or use vagrant)
	if [ -d /vagrant ]; then
		ln -s /vagrant "/opt/$PROJECT"
	else 
		cd /opt; clone $REPO
	fi
	
	# Change apache 2 web document root 
	cat > /etc/apache2/sites-available/$PROJECT.apache.conf <<EOF
	<VirtualHost *:80>
		ServerAdmin webmaster@localhost
		DocumentRoot $WEB_ROOT
		ErrorLog ${APACHE_LOG_DIR}/error.log
		CustomLog ${APACHE_LOG_DIR}/access.log combined
	
	 <Directory $WEB_ROOT >
	    Options "FollowSymLinks"
	    AllowOverride All
	    Require all granted
	    Order allow,deny
	    Allow from all
	  </Directory>
	</VirtualHost>
EOF
	a2dissite 000-default.conf
	a2ensite "$PROJECT.apache.conf"
	service apache2 restart

fi

if [ ! -f "/opt/$PROJECT/composer.lock" ]; then 
	cd "/opt/$PROJECT"; composer install
fi

test/unit/library/BOTK/Context/Context/ConfigTest.php
<?php
namespace BOTK\Context;

use PHPUnit_Framework_TestCase;
use BOTK\Context\ContextNameSpace;
use BOTK\Context\Context as CX;



class ConfigTest extends PHPUnit_Framework_TestCase
{
    public static function setUpBeforeClass()
    {
        // create a temp ini file
        file_put_contents ('/tmp/sample.ini','
; This is a comment
; Section shoud be ignored
[THIS_IS_A_SECTION]
var1 = "This is a string"
var2 = 2 ; an iteger
var3[] = 1  ; array
var3[] = 2  ; array

[ANOTHER_SECTION]
; blank lines ignored
var3[] = 3  ; array

');
        $_ENV['BOTK_CONFIGDIR'] = '/tmp';
    }
    
    public static  function tearDownAfterClass()
    {
        // delete temp ini file
        unlink('/tmp/sample.ini');	
    }
    
    
    public function testVarDefault() {
        $v = CX::factory()->ns('sample')->getValue('novar','ok');
        $this->assertEquals($v,'ok');
    }
    
    
    public function testVarDefaultMandatoryExists() {
    	//die(print_r(CX::factory()->ns('sample'),true));
        $v = CX::factory()->ns('sample')->getValue('var1');
        $this->assertEquals($v,"This is a string");
    }

    
    
    public function testVarArray() {
        $v = CX::factory()->ns('sample')->getValue('var3');
        $this->assertEquals($v,array(1,2,3));
    }

   
     /**
     * @expectedException \Exception
     * 
     */      
    public function testVarDefaultMandatoryNo() {
        $v = CX::factory()->ns('sample')->getValue('novar');
    }


 
    public function testNoNamespace() {
        $v = CX::factory()->ns('nosample')->getValue('novar',1);
        $this->assertEquals($v,1);
    }      
}


test/unit/library/BOTK/Context/ContextNameSpace/ValidationFiltersTest.php
<?php
namespace BOTK\Context;

use PHPUnit_Framework_TestCase;
use BOTK\Context\ContextNameSpace as V;

class ValidatorStubOK
{
    function assert($data) { return true;}
}

class ValidatorStubKO
{
    function assert($data) { throw new \Exception("Error Processing Request", 1);}
}


/**
 * @covers BOTK\Context\ContextNameSpace
 */
class ValidationFiltersTest extends PHPUnit_Framework_TestCase
{
    public $ns;

    public function setUp()
    {
        $this->ns = new ContextNameSpace( array( 
            'p1'    => 'stringa',
            'p2'    => 10,
            'p3'    => array(1,2,3),    // same type array
            'p4'    => array(1,2,'ff'), // mixed type array ( unsupported by std filter, need custom validator)
            'p5'    => new \stdClass,   // unsupported both by ini file and ContextNameSpace, but...
        ));
    }

    public function testDefaultValidator()
    {
        $p = $this->ns->getValue('p1');
        $this->assertEquals($p,'stringa');
    }


    public function testDefaultValidatorWithArray()
    {
        $p = $this->ns->getValue('p4');
        $this->assertEquals($p,array(1,2,'ff'));
    }
	
	
    /**
     * @expectedException \Exception
     */   
    public function testDefaultValidatorWithUnsupportedTypeKO()
    {
        $p = $this->ns->getValue('p5',null,FILTER_VALIDATE_INT);
    }
 
  
    public function testENUM()
    {
         $p = $this->ns->getValue('p1',null, V::ENUM('stringa|string'));
         $this->assertEquals($p,'stringa');       
    }
  
     /**
     * @expectedException \Exception
     */    
    public function testENUMKO()
    {
         $p = $this->ns->getValue('p1',null, V::ENUM('stringx|string'));
    }


    public function testWithCustomValidator()
    {
         $validator = new ValidatorStubOK;
         $p = $this->ns->getValue('p1',null, $validator );
         $this->assertEquals($p,'stringa');       
    }
    
    
     /**
     * @expectedException \Exception
     */  
    public function testWithCustomValidatorKO()
    {
         $validator = new ValidatorStubKO;
         $p = $this->ns->getValue('p1',null, $validator );      
    }
 
    
    public function testSimpleFilterValidator()
    {
        $p = $this->ns->getValue('p2',null, FILTER_VALIDATE_INT);
        $this->assertEquals($p,10);
    }
    
     /**
     * @expectedException \Exception
     */    
    public function testSimpleFilterValidatorKO()
    {
        $p = $this->ns->getValue('p1',null, FILTER_VALIDATE_INT);
    }


    public function testValidatorWithOptions()
    {
        $p = $this->ns->getValue('p2',null,
            array('filter'=>FILTER_VALIDATE_INT,'options'=> array('min_range' => 5, 'max_range' => 15)));
        $this->assertEquals($p,10);
    }
    
    
    /**
     * @expectedException \Exception
     */  
    public function testValidatorWithOptionsKO()
    {
        $p = $this->ns->getValue('p2',null,
            array('filter'=>FILTER_VALIDATE_INT,'options'=> array('min_range' => 11, 'max_range' => 15)));

    }
    /**
     * @expectedException \Exception
     * 
     * Unsupported
     */  
    public function testSimpleFilterValidatorArray()
    {
        $p = $this->ns->getValue('p3',null, FILTER_VALIDATE_INT);
        //$this->assertEquals($p,array(1,2,3));
    }

    /**
     * @expectedException \Exception
     * 
     * Unsupported
     */   
    public function testValidatorWithOptionsArray()
    {
        $p = $this->ns->getValue('p3',null,
            array('filter'=>FILTER_VALIDATE_INT,'options'=> array('min_range' => 0, 'max_range' => 5)));
        //$this->assertEquals($p,array(1,2,3));
    }


    public function testOptionalPar()
    {
        $p = $this->ns->getValue('pnotexists',20 ,FILTER_VALIDATE_INT);
        $this->assertEquals($p,20);
    }
      

 
    public function testNullAsDefault()
    {
        $p = $this->ns->getValue('pnotexists',V::NULL_AS_DEFAULT);
        $this->assertNull($p);
    }


    /**
     * @expectedException \InvalidArgumentException
     * @expectedExceptionMessage Missing mandatory 
     */    
    public function testMandatoryPar()
    {
        $p1 = $this->ns->getValue('pnotexists',null);
    }

}


library/BOTK/Context/ContextNameSpace.php
<?php
namespace BOTK\Context;

use \InvalidArgumentException;

/**
 * Sanitize and validate a variables 
 * 
 * Usage example:
 *  $serverNameSpace = Context::factory()->ns(INPUT_SERVER);
 *  $ip     = $serverNameSpace->getValue('SERVER_ADDR', Context::MANDATORY, FILTER_VALIDATE_IP);
 *  $port   = $serverNameSpace->getValue('SERVER_PORT', '8080', array (
 *      'filter' => FILTER_VALIDATE_INT,
 *      'flags'  => FILTER_REQUIRE_SCALAR,
 *      'options'   => array('min_range' => 8080, 'max_range' => 8084)
 *  )
 * 
 * You can also use Respect\Validation library for complex validator:
 *  $port = $serverNameSpace->getValue('SERVER_PORT', '8080',v::int()->between(8080,8084))
 * 
 * You can validate and sanitize in one call:
 * 
 *  $port = $serverNameSpace->getValue('SERVER_PORT', '8080', v::int()->between(8080,8084),FILTER_SANITIZE_NUMBER_INT);
 * 
 * You can use Context helper hortcuts:
 *  $method = $serverNameSpace->getValue('REQUEST_METHOD', 'GET', $c->ENUM('GET|POST'))
 * 
 * Note that reurned value can be a scalar or an array.
 */
class ContextNameSpace
{
    const MANDATORY = null;                 // just syntactic sugar for null value
    const NULL_AS_DEFAULT = '__null__';		// a string that allow null to be used as a default falue (by default null means no default)
       
    protected $varStore = array();
    
    /**
     * 
     * @param $theArray can be an array or a reference to a predefinite array (es. $_POST)
     */
    public function __construct( $theArray )
    {
        $this->varStore = $theArray;
    }
    
    
    /**
     * Sanitize and validate a variable contained in a name space
     *
     * @param string $varName       the name of a variable defined in the namespace
     * @param mixed  $default       a default value if variable is not set. NULL means that
     *                              the variable is mandatory.
     * @param mixed  $validator     Can be:
     *                                1) null use default validation
     *                                2) an integer representing a simple Validate filter without flags and options
     *                                3) an array representing a Validate filter with flags and/or options
     *                                4) an instance of an object that expose function 'assert'.
     *                                   you can use istance of Respect\Validation\Validator

     * @param mixed  $sanitizer     Can be:
     *                                A) null if no sanitization is required
     *                                B) an integer representing a simple Sanitize filter without flags and options
     *                                C) an array representing a Sanitize filter with flags and/or options
     *                                D) a callable that accept an argument (the source) and return sanitized one 
     * 
     * @throws InvalidArgumentException if $varName is not defined and $default not provided
     * @throws Exception if validator fails to validate variable (depending from the validator)
     * @throws InvalidArgumentException if sanitize fails
     * 
     * @return mixed the sanitized value associated to varName in namespace
     * 
     * LIMITS:
     *  sanitizer are supported just on scalar values
     *  non scalar value require a custom validator like Respect\Validator
     * 
     */
    public function getValue (
        $varName,
        $default = null,
        $validator=null,
        $sanitizer=null
    )
    {
        // prepare a static array to validate validators and sanitizer filters   
        static $validFilters = null;  //load once runtime
        if (is_null($validFilters)){
            // prepare a list of valid filters: do once 
            $validFilters= array();
            $filters = filter_list(); 
            foreach($filters as $filter_name) { 
                $validFilters[filter_id($filter_name)] = $filter_name; 
            }
            // Remove unsupported filters:
            foreach( array(FILTER_VALIDATE_BOOLEAN,FILTER_CALLBACK ) as $badFilter){
                unset($validFilters[$badFilter]);
            }
        }
        
        // Ensure var exists or a default value is present
        if (!$varName
            || (is_null($default) && !isset($this->varStore[$varName])))  {
            throw new InvalidArgumentException("Missing mandatory parameter $varName.",400);                
        }
     
        $sourceValue = isset($this->varStore[$varName])?$this->varStore[$varName]:$default;
        $scalarValue = is_scalar($sourceValue);

		// manage NULL_AS_DEFAULT
		if($sourceValue===self::NULL_AS_DEFAULT) {$sourceValue=null;}
        
        // sanitize the variable value (if requested )
        if (is_null($sanitizer)) {
            // CASE A: sanitizing not required
            $sanitizedValue = $sourceValue;
        } elseif (is_int($sanitizer) && array_key_exists($sanitizer,$validFilters) && $scalarValue) {          
            // CASE B: Use a simple sanitize filter
            $sanitizedValue = filter_var($sourceValue, $sanitizer);
        } elseif (is_array($sanitizer) 
            && isset($sanitizer['filter']) 
            && array_key_exists($sanitizer['filter'],$validFilters)
            && $scalarValue
        ) {
            // CASE C: use sanitizer filter with flags and/or options
            $filter = $sanitizer['filter'];
            $sanitizedValue = filter_var($sourceValue, $filter, $sanitize); // should I unset($sanitizer['filter']) ?
        } elseif (is_callable($sanitizer)) {
            $sanitizedValue = $sanitizer($sourceValue);
        } else {
            throw new InvalidArgumentException("Invalid sanitizer filter on $varName.",400);            
        }
 
        //Now we are sure that variable has a sanitized value: validate it (if requested )
        
        if (is_null($validator)) {
            // CASE 1: validation not required
            $valid = $scalarValue?filter_var($sanitizedValue):$sanitizedValue;
            $validatorName='DEFAULT';
        } elseif (is_int($validator) 
            && array_key_exists($validator,$validFilters)
            && $scalarValue ) {
            // CASE 2: Use a simple validation filter
            $valid = filter_var($sanitizedValue, $validator);
            $validatorName=$validFilters[$validator];
        } elseif (is_array($validator) 
            && isset($validator['filter']) 
            && array_key_exists($validator['filter'],$validFilters)
            && $scalarValue ){
            // CASE 3: use validation filter with flags and/or options
            $filter = $validator['filter'];
            $valid = filter_var($sanitizedValue, $filter, $validator); // should unset($validator['filter'])?
            $validatorName=$validFilters[$filter];
        } elseif (is_object($validator)) {
            // CASE 4: the object must expose assert() method (like in Respect\Validation library )
            try {
                $validator->assert($sanitizedValue);  // N.B. is up to custom validator test the value type
                $valid=true;
            } catch ( Exception $e){
                $validatorName='OBJECT_VALIDATOR';
                $valid = false;
                $errorMsg=$e->getMessage();
            }
        } else {
            $validatorName='unsupported';
            $valid=false;
        }
        
        // thrown exception if not validate value
        if (false===$valid) {
            if (empty($errorMsg)) {
                $val=$scalarValue?$sanitizedValue:'structured';
                $errorMsg = "Invalid value($val) for $validatorName filter on [$varName]";
            }
            throw new InvalidArgumentException($errorMsg,400);
        }
        
        return $sanitizedValue;
    }

    /*
     * Here are some helpers for  shortcuts
     */
    public function getString($varName,$default = null)
    {
        return $this->getValue($varName,$default, null, FILTER_SANITIZE_STRING);
    }
    
    public function getURI($varName,$default = null)
    {
        return $this->getValue($varName,$default, self::STRING('/.+/'), FILTER_SANITIZE_URL);
    }


         
    /*
     * Here are some helpers for validator shortcuts
     */
     public static function ENUM($wlist)
     {
         return  array(
                'filter'    => FILTER_VALIDATE_REGEXP,
                'options'   => array('regexp' => "/^($wlist)$/")
         );
     }
     
     public static function STRING($pattern)
     {
         return  array(
                'filter'    => FILTER_VALIDATE_REGEXP,
                'options'   => array('regexp' => $pattern)
         );
     }

     public static function FILENAME()
     {
         return  array(
                'filter'    => FILTER_VALIDATE_REGEXP,
                'options'   => array('regexp' => '/^[\w,\s-]+\.[A-Za-z]+$/')
         );
     }  


     public static function POSITIVE_INT()
     {
         return  array(
                'filter'    => FILTER_VALIDATE_INT,
                'options'   => array("min_range"=>1)
         );
     }


     public static function NON_NEGATIVE_INT()
     {
         return  array(
                'filter'    => FILTER_VALIDATE_INT,
                'options'   => array("min_range"=>0)
         );
     }  
}


library/BOTK/Context/Context.php
<?php
namespace BOTK\Context;

use BOTK\Context\ContextNameSpace;


class Context
{
    const LOCAL = 'LOCAL_VARS';
    
    
    protected $configDir = '../configs';    // default place where to search for ini file
    protected $nameSpaces = array();        // contains namespaces

    
    public static function factory(array $localVars = array())
    {
        return new static($localVars);
    }
    
    
    public function __construct( array $localVars = array())
    {
        // register predefined name space
        $this->nameSpaces = array(
            INPUT_POST      => new ContextNameSpace($_POST),
            INPUT_GET       => new ContextNameSpace($_GET),
            INPUT_ENV       => new ContextNameSpace($_ENV),
            INPUT_SERVER    => new ContextNameSpace($_SERVER),
            self::LOCAL     => new ContextNameSpace($localVars)
            // This is supposed a RESTfull toolkit ... no COOKIE, no SESSION no GLOBALS by design :-) 
        );
        
        // initialize external namespace technology
        $this->configureIniStore();         // this is for namespace maintained in ini file
    }
    
    
    /*
     * Where to find config files
     * Fall back priorities
     *  case 1) use the one provided as paramether
     *  case 2) check the one provided in BOTK_CONFIGDIR environment variable
     *  case 3) use ../configs relative to application script file
     */
    public function configureIniStore( $configDir=null)
    {
        $this->configDir = $configDir                                           // case 1)
            ? $configDir
            : ( isset($_ENV['BOTK_CONFIGDIR'])                                    // case 2
                ? $_ENV['BOTK_CONFIGDIR'] 
                : (dirname(dirname($_SERVER['SCRIPT_FILENAME'])) . '/configs')
               ); // case 3
                
        return $this;
    }
    

    private function loadArrayFromIniFile($name)
    {
        static $inifilesDB = array();   // a database of variables loaded from .ini files
        
        if (!isset($inifilesDB[$name])) {
            // Let error management take care of errors...
            $iniFile = "$this->configDir/$name.ini";
            $inifilesDB[$name]= file_exists($iniFile)
                ?parse_ini_file($iniFile)
                :array();
            if (!is_array($inifilesDB[$name])){
                throw new \InvalidArgumentException("Invalid configuration file in $iniFile.",404);                
            }            
        }
        
        return $inifilesDB[$name];
    }
    
    /*
     * $nameSpace can be:
     *  . a predefined name space idenfifier (INPUT_POST, INPUT_GET, INPUT_ENV,INPUT_SERVER )
     *  . a string containing a file name that mst exists in configDir with .ini extension
     * 
     */
    public function ns($nameSpace)
    {
        // If ! exist try to load a configuration file 
        if ( !isset($this->nameSpaces[$nameSpace]) ) {
            // short sanitize of nameSpace
            if (!filter_var( $nameSpace,FILTER_VALIDATE_REGEXP, 
                array('options'=>array('regexp' => '/^[\w]+$/')))) {
                    throw new InvalidArgumentException("Invalid configuration name.",400);
            }
            $this->nameSpaces[$nameSpace] = new ContextNameSpace($this->loadArrayFromIniFile($nameSpace));
        }
        
        return $this->nameSpaces[$nameSpace];
    } 
    
   /*
     * This helper tryes to return the canonical uri of current http requests.
     * It is inspired by Joomla JURI:base() implementation
     * 
     * Tested with apache server.
     */
    public function guessRequestCanonicalUri()
    {
        $server = $this->ns(INPUT_SERVER);
        
        // Determine if the request was over SSL (HTTPS).
        if( (strtolower($server->getValue('HTTPS','off')) != 'off')) {
            $schema = 'https://';
            $defaultPort=443;
        } else {
            $schema = 'http://';
            $defaultPort=80;           
        }
        
        $port = ($server->getValue('SERVER_PORT',$defaultPort, ContextNameSpace::NON_NEGATIVE_INT()));
        $http_host = $server->getValue('HTTP_HOST','localhost');        
        $authority = ($port===$defaultPort)
            ?($http_host)
            :($http_host . ':'. $port);
                
        return $schema.$authority.$this->guessRequestRelativeUri();
    }


   /*
     * This helper tryes to return the relative uri of current http requests.
     * We need this since in IIS we do not have REQUEST_URI to work with, we will assume we are
     * and so will therefore need to work some magic with the SCRIPT_NAME 
     * 
     * Tested with apache server.
     */
    public function guessRequestRelativeUri()
    {
        $server = $this->ns(INPUT_SERVER);
        
        $php_self = $server->getValue('PHP_SELF','');
        $request_uri = $server->getValue('REQUEST_URI','');
         
        if (!empty($php_self) && !empty($request_uri))
        {
            $theURI = $request_uri;
        }
        else
        {
            // Since we do not have REQUEST_URI to work with, we will assume we are
            // running on IIS and will therefore need to work some magic with the SCRIPT_NAME and
            // QUERY_STRING environment variables.
            $script_name = $server->getValue('SCRIPT_NAME','');
            $query_string = $server->getValue('QUERY_STRING','');
            

            // IIS uses the SCRIPT_NAME variable instead of a REQUEST_URI variable... thanks, MS
            $theURI = $script_name;

            // If the query string exists append it to the URI string
            if (!empty($query_string))
            {
                $theURI .= '?' . $query_string;
            }
        }
        
        return $theURI;
    }
}
